webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/TextValue/TextValuePresentation.tsx

```diff
@@ -70,82 +69,72 @@ const styles = css`
-    let readonly = true;
-    let valueTruncated = false;
-    let limit: string | undefined;
-    let valueSize: string | undefined;
-
-    if (activeElements.length > 0) {
-      const format = model.source.getAction(resultIndex, ResultSetFormatAction);
-
-      firstSelectedCell = activeElements[0];
-
-      const value = format.get(firstSelectedCell);
-      readonly = format.isReadOnly(firstSelectedCell) || format.isBinary(firstSelectedCell);
-
-      if (isResultSetContentValue(value)) {
-        valueTruncated = content.isContentTruncated(value);
-
-        if (valueTruncated) {
-          limit = bytesToSize(quotasService.getQuota('sqlBinaryPreviewMaxLength'));
-          valueSize = bytesToSize(value.contentLength ?? 0);
-        }
-
-        if (value.contentType) {
-          contentType = value.contentType;
+    const firstSelectedCell = activeElements?.[0];
+    const activeTabs = textValuePresentationService.tabs.getDisplayed({
+      dataFormat: dataFormat,
+      model: model,
+      resultIndex: resultIndex,
+    });
+    const { contentAction, editAction, formatAction } = useResultActions({
+      model,
+      resultIndex,
+    });
+    const contentValue = firstSelectedCell ? formatAction.get(firstSelectedCell) : null;
+    const state = useTabLocalState(() =>
+      observable({
+        currentContentType: TEXT_PLAIN_TYPE,
 
-          if (contentType === 'text/json') {
-            contentType = 'application/json';
+        setContentType(contentType: string) {
+          if (contentType === TEXT_JSON_TYPE) {
+            contentType = APPLICATION_JSON_TYPE;
           }
 
           if (!activeTabs.some(tab => tab.key === contentType)) {
-            contentType = 'text/plain';
+            contentType = TEXT_PLAIN_TYPE;
           }
-        }
-      }
-    }
 
-    readonly = model.isReadonly(resultIndex) || model.isDisabled(resultIndex) || readonly;
-
-    if (contentType !== state.lastContentType) {
-      state.setDefaultContentType(contentType);
-    }
+          this.currentContentType = contentType;
+        },
+        handleTabOpen(tabId: string) {
+          // currentContentType may be selected automatically we don't want to change state in this case
+          if (tabId !== state.currentContentType) {
+            state.setContentType(tabId);
+          }
+        },
+      }),
+    );
+    const { textValue, isFullTextValue, isTruncated, isTextColumn } = useTextValue({
+      model,
+      resultIndex,
+      currentContentType: state.currentContentType,
+    });
+    const isSelectedCellReadonly = firstSelectedCell && (formatAction.isReadOnly(firstSelectedCell) || formatAction.isBinary(firstSelectedCell));
+    const isReadonlyByResultIndex = model.isReadonly(resultIndex) || model.isDisabled(resultIndex) || !firstSelectedCell;
+    const isReadonly = isSelectedCellReadonly || isReadonlyByResultIndex;
+    const valueSize = bytesToSize((contentValue as IResultSetContentValue)?.contentLength ?? 0);
```

Comment:
it's better to use `isResultSetContentValue` to ensure that value is `IResultSetContentValue`